File: sih26168_v2x/python/sih26168_v2x/security.py
from __future__ import annotations

from .types import NormalizedV2XMessage, SecurityStatus, V2XConfig


class MockSecurityProvider:
    """No cryptography. Never authenticates. MODE A simulation only."""

    def evaluate(self, msg: NormalizedV2XMessage, cfg: V2XConfig) -> tuple[SecurityStatus, str]:
        dec = msg.declared_security
        if dec in (SecurityStatus.REJECTED, SecurityStatus.INVALID, SecurityStatus.UNTRUSTED):
            if cfg.reject_untrusted:
                return SecurityStatus.REJECTED, "declared_untrusted"
            return SecurityStatus.UNTRUSTED, "untrusted_allowed"
        if dec == SecurityStatus.AUTHENTICATED:
            return SecurityStatus.UNVERIFIED, "claimed_auth_ignored"
        if cfg.reject_unverified:
            return SecurityStatus.REJECTED, "unverified_rejected"
        return SecurityStatus.UNVERIFIED, "unverified_mock"

File: sih26168_v2x/src/security.cpp
#include "sih26168/v2x/security.hpp"

namespace sih26168::v2x {

SecurityStatus MockSecurityProvider::evaluate(const NormalizedV2XMessage& msg, const V2XConfig& cfg,
                                              std::string& reason) const {
    // MODE A: no cryptography. Never upgrades a message to Authenticated.
    switch (msg.declared_security) {
        case SecurityStatus::Rejected:
        case SecurityStatus::Invalid:
        case SecurityStatus::Untrusted:
            if (cfg.reject_untrusted) {
                reason = "declared_untrusted";
                return SecurityStatus::Rejected;
            }
            reason = "untrusted_allowed";
            return SecurityStatus::Untrusted;
        case SecurityStatus::Authenticated:
            if (cfg.allow_claimed_authenticated_simulation) {
                reason = "claimed_authenticated_simulation_only";
                return SecurityStatus::Unverified;
            }
            reason = "claimed_auth_ignored";
            return SecurityStatus::Unverified;
        case SecurityStatus::Unverified:
        case SecurityStatus::Received:
        case SecurityStatus::Validated:
        case SecurityStatus::Stale:
        default:
            if (cfg.reject_unverified) {
                reason = "unverified_rejected";
                return SecurityStatus::Rejected;
            }
            reason = "unverified_mock";
            return SecurityStatus::Unverified;
    }
}

std::unique_ptr<IV2XSecurityProvider> makeMockSecurityProvider() {
    return std::make_unique<MockSecurityProvider>();
}

}  // namespace sih26168::v2x
